desc/compute/_neoclassical.py

```diff
@@ -0,0 +1,221 @@
+expensive computations.
+"""
+
+from functools import partial
+
+from quadax import simpson
+
+from desc.backend import jit, jnp
+
+from ..integrals.bounce_integral import Bounce1D
+from ..integrals.bounce_utils import get_pitch_inv_quad
+from ..integrals.quad_utils import chebgauss2
+from ..utils import map2, safediv
+from .data_index import register_compute_fun
+
+
+def _alpha_mean(f):
+    """Simple mean over field lines.
+
+    Simple mean rather than integrating over α and dividing by 2π
+    (i.e. f.T.dot(dα) / dα.sum()), because when the toroidal angle extends
+    beyond one transit we need to weight all field lines uniformly, regardless
+    of their spacing wrt α.
+    """
+    return f.mean(axis=0)
+
+
+def _get_pitch_inv_quad(grid, data, num_pitch, _data):
+    p, w = get_pitch_inv_quad(
+        grid.compress(data["min_tz |B|"]), grid.compress(data["max_tz |B|"]), num_pitch
+    )
+    _data["pitch_inv"] = jnp.broadcast_to(
+        p[jnp.newaxis], (grid.num_alpha, grid.num_rho, num_pitch)
+    )
+    _data["pitch_inv weight"] = jnp.broadcast_to(
+        w[jnp.newaxis], (grid.num_alpha, grid.num_rho, num_pitch)
+    )
+    return _data
+
+
+@register_compute_fun(
+    name="<L|r,a>",
+    label="\\int_{\\zeta_{\\mathrm{min}}}^{\\zeta_{\\mathrm{max}}}"
+    " \\frac{d\\zeta}{|B^{\\zeta}|}",
+    units="m / T",
+    units_long="Meter / tesla",
+    description="(Mean) proper length of field line(s)",
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=["B^zeta"],
+    resolution_requirement="z",
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+)
+def _L_ra_fsa(data, transforms, profiles, **kwargs):
+    grid = transforms["grid"].source_grid
+    L_ra = simpson(
+        y=grid.meshgrid_reshape(1 / data["B^zeta"], "arz"),
+        x=grid.compress(grid.nodes[:, 2], surface_label="zeta"),
+        axis=-1,
+    )
+    data["<L|r,a>"] = grid.expand(jnp.abs(_alpha_mean(L_ra)))
+    return data
+
+
+@register_compute_fun(
+    name="<G|r,a>",
+    label="\\int_{\\zeta_{\\mathrm{min}}}^{\\zeta_{\\mathrm{max}}}"
+    " \\frac{d\\zeta}{|B^{\\zeta} \\sqrt g|}",
+    units="1 / Wb",
+    units_long="Inverse webers",
+    description="(Mean) proper length over volume of field line(s)",
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=["B^zeta", "sqrt(g)"],
+    resolution_requirement="z",
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+)
+def _G_ra_fsa(data, transforms, profiles, **kwargs):
+    grid = transforms["grid"].source_grid
+    G_ra = simpson(
+        y=grid.meshgrid_reshape(1 / (data["B^zeta"] * data["sqrt(g)"]), "arz"),
+        x=grid.compress(grid.nodes[:, 2], surface_label="zeta"),
+        axis=-1,
+    )
+    data["<G|r,a>"] = grid.expand(jnp.abs(_alpha_mean(G_ra)))
+    return data
+
+
+@register_compute_fun(
+    name="effective ripple",  # this is ε¹ᐧ⁵
```

Comment:
do you have the math where you cancelled out the intermediate B0 etc from nemov's paper somewhere?

Reply:
well it's in a notebook back in texas. B_0 is a constant you can always take out though, and b' is a constant you can take out of stuff up until the integral over db'. then the comments have the math for converting the integral